### PR TITLE
Hhvm strict hackarrays

### LINKED ISSUE #293: Use hack arrays over collections
Use hack arrays over collections, into [hhvm_strict](https://github.com/facebook/facebook-instant-articles-sdk-php/tree/hhvm_strict) branch.

### PR DESCRIPTION
**Pull request by everton-rosario**
This PR

* [x] Replaced all Map => dict
* [x] Replaced all Vector => vec 
* [x] Replaced all Set => keyset
* [x] Replaced all array => dict/vec/keyset (using the proper structure)

Fixes #293

### CODE REVIEW

File: tests/Facebook/InstantArticles/Elements/Validators/InstantArticleValidatorTest.php [hunk 1]
```diff
@@ -113,6 +113,12 @@ public function testInstantArticle(): void
         $this->assertEquals($expected, $result);
 
         $warnings = InstantArticleValidator::check($article);
+
+        foreach($warnings as $warning) {
+          var_dump($warning->__toString());
+        }
```
Comment (mburak) [inline tests/Facebook/InstantArticles/Elements/Validators/InstantArticleValidatorTest.php]: I guess you can remove this logging.

Comment (mburak) [reply tests/Facebook/InstantArticles/Elements/Validators/InstantArticleValidatorTest.php]: I'm sorry, this looks to be old code.

Comment (everton-rosario) [reply tests/Facebook/InstantArticles/Elements/Validators/InstantArticleValidatorTest.php]: No problem, dropping it!
Thanks for flagging!

File: src/Facebook/InstantArticles/Validators/InstantArticleValidator.php [hunk 2]
```diff
@@ -24,29 +24,28 @@ class InstantArticleValidator
...
     {
         foreach ($elements as $element) {
             if (!$element->isValid() && $element->isEmptyValidationEnabled()) {
                 // Adds a warning to the result report.
-                $warnings->add(new ValidatorWarning($element));
+                $warnings[] = new ValidatorWarning($element);
             }
             if ($element instanceof ChildrenContainer) {
-                self::getReport($element->getContainerChildren(), $warnings);
+                $warnings = Type::concatVec($warnings, self::getReport($element->getContainerChildren(), vec[]));
```
Comment (mburak) [inline src/Facebook/InstantArticles/Validators/InstantArticleValidator.php]: Can't you use the existing Vec\concat() function here?

Comment (everton-rosario) [reply src/Facebook/InstantArticles/Validators/InstantArticleValidator.php]: First run was with few issues then I dropped and forgot to re-add it. Good call.
Pushed.

File: src/Facebook/InstantArticles/Transformer/Rules/Rule.php [hunk 3]
```diff
@@ -53,20 +39,20 @@ public static function create(): Rule
...
     }
 
-    public static function retrieveProperty(array<string, mixed> $properties, string $property_name): ?array<string, mixed>
+    public static function retrieveProperty(dict<string, mixed> $properties, string $property_name): ?dict<string, mixed>
     {
         if (array_key_exists('properties', $properties)) {
             $mappedProperties = $properties['properties'];
             if (is_array($mappedProperties) && array_key_exists($property_name, $mappedProperties)) {
-                return $mappedProperties[$property_name];
+                return dict($mappedProperties[$property_name]);
```
Comment (diegoquinteiro) [inline src/Facebook/InstantArticles/Transformer/Rules/Rule.php]: What if `$mappedProperties[$property_name]` is not an array (bogus rules JSON)?

Example:

```json
{ "rules":[{"class":"ImageRule", "selector": "div.image", "properties": "clowntown"}] }
```

File: src/Facebook/InstantArticles/Validators/Type.php [hunk 4]
```diff
@@ -44,121 +44,121 @@ private static function throwException(mixed $var, mixed $types_allowed): void
...
      *
-     * @param array $array the array that will be checked
-     * @param int $min_size The minimum elements the array must have
+     * @param vec $vec the vec that will be checked
+     * @param int $min_size The minimum elements the vec must have
      *
      * @return bool true if has at least $min_size, false otherwise
      */
-    public static function isArraySizeGreaterThan(Vector<mixed> $array, int $min_size, bool $enforce = false): bool
+    public static function isArraySizeGreaterThan(vec<mixed> $vec, int $min_size, bool $enforce = false): bool
```
Comment (diegoquinteiro) [inline src/Facebook/InstantArticles/Validators/Type.php]: Maybe rename to `isVecSizeGreaterThan()`

File: src/Facebook/InstantArticles/Validators/Type.php [hunk 5]
```diff
@@ -44,121 +44,121 @@ private static function throwException(mixed $var, mixed $types_allowed): void
...
+     * @param vec $vec the vec that will be checked
+     * @param int $max_size The maximum number of items the vec can have
      *
      * @return bool
      *
-     * @throws \InvalidArgumentException if $array have more than $max_size items
+     * @throws \InvalidArgumentException if $vec have more than $max_size items
      */
-    public static function enforceArraySizeLowerThan(Vector<mixed> $array, int $max_size): bool
+    public static function enforceArraySizeLowerThan(vec<mixed> $vec, int $max_size): bool
```
Comment (diegoquinteiro) [inline src/Facebook/InstantArticles/Validators/Type.php]: Same here, let's use Vec on the names.

File: src/Facebook/InstantArticles/Validators/Type.php [hunk 6]
```diff
@@ -44,121 +44,121 @@ private static function throwException(mixed $var, mixed $types_allowed): void
...
-     * @param int $max_size The maximum number of items the array can have
+     * @param vec $vec the vec that will be checked
+     * @param int $max_size The maximum number of items the vec can have
      * @param boolean $enforce works as Type::enforceArrayMaxSize().
      * @see Type::enforceArrayMaxSize().
      *
      * @return bool true if it has less elements than $max_size, false otherwise
      */
-    public static function isArraySizeLowerThan(Vector<mixed> $array, int $max_size, bool $enforce = false): bool
+    public static function isArraySizeLowerThan(vec<mixed> $vec, int $max_size, bool $enforce = false): bool
```
Comment (diegoquinteiro) [inline src/Facebook/InstantArticles/Validators/Type.php]: Same here =)

File: src/Facebook/InstantArticles/Validators/Type.php [hunk 7]
```diff
@@ -44,121 +44,121 @@ private static function throwException(mixed $var, mixed $types_allowed): void
...
         }
         return $meets_size;
     }
 
     /*
-     * Utility method that constructs the message about array sizes an throws.
+     * Utility method that constructs the message about vec sizes an throws.
      */
-    private static function throwArrayException(Vector<mixed> $array, int $size, string $message): void
+    private static function throwArrayException(vec<mixed> $vec, int $size, string $message): void
```
Comment (diegoquinteiro) [inline src/Facebook/InstantArticles/Validators/Type.php]: Same here.

File: tests/Facebook/InstantArticles/Transformer/LoaderRulesTest.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+use Facebook\InstantArticles\Elements\Header;
+use Facebook\InstantArticles\Transformer\Rules\ConfigurationSelectorRule;
+use Facebook\InstantArticles\Transformer\Rules\BoldRule;
+use Facebook\InstantArticles\Transformer\Rules\H1Rule;
+use Facebook\InstantArticles\Transformer\Rules\ItalicRule;
+use Facebook\InstantArticles\Transformer\Rules\ParagraphRule;
+use Facebook\InstantArticles\Transformer\Rules\TextNodeRule;
+
+class LoaderRulesTest extends \Facebook\Util\BaseHTMLTestCase
```
Comment (diegoquinteiro) [inline tests/Facebook/InstantArticles/Transformer/LoaderRulesTest.php]: Nice 💅

File: src/Facebook/InstantArticles/Transformer/Rules/Rule.php (hunk 3 above)
Comment (everton-rosario) [reply src/Facebook/InstantArticles/Transformer/Rules/Rule.php]: This is all fine, added few more Test cases that raises no error while processing.

**Pull request merged**